Quest Manager refactoring

magicus opened the pull request:
In preparation of the upcoming /quest command, refactor QuestBookManager: 
 * Rename to QuestManager
 * Collect all quest related functionality under one roof
 * Split up the unwieldy QuestManager in several, simpler and more specialized classes

magicus commented:
This fix screams for generalisation. :-) I won't force you, but I do think it'd be better off if you introduced a method like `getSearchString(ItemStack is, boolean isLore)` or something like that. Then read and strip "lore": and use it to set isLore. And just let this function return `ItemUtils.getStringLore(is))` or `is.getDescription()` depending on the value.

Incompleteusern commented:
LGTM. A potential extension could be using ItemsPage's advanced searchbar code and adding the lore section to it, generalization is possible.

DonkeyBlaster69 commented:
> This fix screams for generalisation. :-) I won't force you, but I do think it'd be better off if you introduced a method like `getSearchString(ItemStack is, boolean isLore)` or something like that. Then read and strip "lore": and use it to set isLore. And just let this function return `ItemUtils.getStringLore(is))` or `is.getDescription()` depending on the value.

Bit confused on what you mean by this, I'm gonna write something up and you can see if it's what you meant?

magicus commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/BankOverlay.java:
```diff
@@ -393,14 +392,31 @@ private void searchPageForItems(ChestReplacer bankGui) {
     }
 
+    private String getSearchString(ItemStack is, boolean isLore) {
+        if (isLore) return TextFormatting.getTextWithoutFormattingCodes(ItemUtils.getStringLore(is).toLowerCase());
+        return TextFormatting.getTextWithoutFormattingCodes(is.getDisplayName().toLowerCase());
```
```suggestion
        if (isLore) {
          return TextFormatting.getTextWithoutFormattingCodes(ItemUtils.getStringLore(is).toLowerCase());
        } else {
          return TextFormatting.getTextWithoutFormattingCodes(is.getDisplayName().toLowerCase());
        }
```

magicus reviewed (approved):
Yes, exactly. Much better! I added a suggestion with code formatting. Apart from that, this looks good to me now.

DonkeyBlaster69 commented on src/main/java/com/wynntils/modules/utilities/overlays/inventories/BankOverlay.java:
```diff
@@ -393,11 +393,24 @@ private void searchPageForItems(ChestReplacer bankGui) {
-            ItemStack is = bankGui.getLowerInv().getStackInSlot(i);
-            if (TextFormatting.getTextWithoutFormattingCodes(is.getDisplayName()).toLowerCase().contains(searchText)) searchedItems.add(is);
+                ItemStack is = bankGui.getLowerInv().getStackInSlot(i);
+                if (TextFormatting.getTextWithoutFormattingCodes(is.getDisplayName()).toLowerCase().contains(searchText))
```
Got it, thanks!

magicus committed:
Initial /quest command implementation

kristofbolyai commented on common/src/main/java/com/wynntils/commands/QuestCommand.java:
```diff
@@ -0,0 +1,154 @@
+
+        QuestInfo quest = matchingQuests.get(0);
+
+        MutableComponent response =
```
It will be so much fun to I18n commands. (This also applies to almost every other command class though..)

magicus replied:
Not sure we need to do that, though. If we consider commands to be "advanced user" thing, it can probably stay English. (Don't know if you ever run like a localized Linux, and just gotten so annoyed at error messages being incomprehensible in your native language instead of English :-))

magicus committed:
Merge branch 'main' into quest-book-stuff

magicus committed:
Fix merge errors

magicus committed:
A mixed bag of updates:
 * Implement /quest wiki, track, untrack commands
 * Sort quests on distance, name, level
 * Move more logic into QuestBookManager and QuestInfo

magicus committed:
chore: spotless formatting

magicus committed:
Rename QuestBookManager to QuestManager (and also change the package name)

magicus committed:
chore: spotless formatting

magicus committed:
Merge branch 'main' into quest-book-stuff

magicus committed:
chore: spotless formatting

magicus committed:
Fold QuestHandler into QuestManager

(Also fix some merge errors)

magicus committed:
chore: spotless formatting

magicus committed:
Refactor: introduce resetHandler in ScoreboardHandler

magicus committed:
Fold ScoreboardQuestInfo into QuestInfo

magicus committed:
chore: spotless formatting

magicus committed:
QuestManager is too large, split up in separate classes

magicus committed:
Deduplicate getQuestBookTitle

magicus committed:
Extract quest info parsing into QuestInfoParser

magicus committed:
chore: spotless formatting

magicus committed:
Skip the /quest command for now, focus on refactoring

magicus committed:
Merge branch 'main' into quest-book-stuff

magicus committed:
Add warning printout

kristofbolyai committed:
Merge branch 'main' into quest-book-stuff

kristofbolyai reviewed (dismissed):
LGTM

Incompleteusern commented:
Still reviewing, but likely is fine

Incompleteusern commented on common/src/main/java/com/wynntils/gui/widgets/QuestButton.java:
```diff
@@ -125,27 +118,8 @@ private void trackQuest() {
-        RequestHandler handler = new RequestHandler();
-
-        handler.addAndDispatch(req, true);
+        McUtils.soundManager().play(SimpleSoundInstance.forUI(SoundEvents.EXPERIENCE_ORB_PICKUP, 1.0F));
```
We should have a helper method that plays sounds, not really this pr however

magicus replied:
It's not that complicated, but I agree, a helper would be a good thing.

https://github.com/Wynntils/Artemis/issues/509

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/DialogueHistoryQueries.java:
```diff
@@ -0,0 +1,99 @@
+        queryBuilder.build().executeQuery();
+    }
+
+    private void createActualDialogueHistoryQuery(int pageCount) {
```
I would suggest removing the `actual` from this, as this is more of a continuation of the query

magicus replied:
Well, it *is* creating the actual query, since the other part is just a helper to extract the number of pages we need to query.

I've renamed it slightly, and added some more clarification, so the flow is hopefully clearer.

Me and boylai need to come up with a better way to create dynamic queries, so we don't need to know the number of pages beforehand. But that is a future exercise.

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/QuestInfo.java:
```diff
@@ -0,0 +1,177 @@
+
+    public String getName() {
+        return name;
+    }
```
Could this reasonably be made a record?

magicus replied:
No, since nextTask is not final (and actually `status` should not be either). If we really wanted to, maybe we should split this into like a constant `QuestInfo` with just title, level, etc, and a `QuestProgress` which updates as we progress through a quest. 

It will help the conceptual design, but it will also make us need more code to compensate for the abstraction. I'll have to think about it. So not this PR at least. (But it's likely not the last one in QuestManager)

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/QuestInfo.java:
```diff
@@ -0,0 +1,177 @@
+
+    public void setNextTask(String nextTask) {
+        this.nextTask = nextTask;
+    }
```
I suggest renaming `task` to `step` as we use `task` already for render tasks

magicus replied:
And we also use "step" for the next step in queries, which is much tighter coupled with the QuestManager than rendering is. 

I've used the term `nextTask` consistently, so it's never called just `task`. 

I think we'll just have to accept that commonplace words like "step" and "task" can be used with different meaning in different contexts.

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/QuestManager.java:
```diff
@@ -0,0 +1,190 @@
+        return sortQuestInfoList(sortOrder, quests);
+    }
+
+    public static List<QuestInfo> getMiniQuestsSorted(QuestSortOrder sortOrder) {
```
Rename to have `Sorted` before `MiniQuests`/`Quests`?

magicus replied:
Actually, this should be the main way to get quests; the "raw" methods are (currently) not even used. I removed those and skipped the "sorted" part. If we ever need raw access again, let's name it something with "raw" instead.

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/QuestManager.java:
```diff
@@ -0,0 +1,190 @@
+
+        // TODO handle mini quest
+        String name = questInfo.getName();
+        String wikiQuestPageNameQuery = WebManager.getApiUrl("WikiQuestQuery");
```
check if apiurls exists before using

magicus replied:
Sigh, the apiurls overhead is just bad. :(

Fixed it. Also noticed that the url pointed to the old domain...

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/QuestManager.java:
```diff
@@ -0,0 +1,190 @@
+            Optional<Location> loc2 = quest2.getNextLocation();
+            return (int) (getDistance(loc1) - getDistance(loc2));
+        }
+    }
```
This can reasonably be extracted to another class

magicus replied:
What do you mean? It is another class already. Or do you mean that it should be a top-level class in its own file? If so, I disagree. This is just a small helper class; it was slightly too large to be a lambda, but it's only used in sortQuestInfoList.

Incompleteusern replied:
I feel that it might be used in other places, but we can move it if that occurs. Indeed, it is only used here for now.

Incompleteusern commented on common/src/main/java/com/wynntils/wynn/model/quests/QuestScoreboardHandler.java:
```diff
@@ -29,35 +23,38 @@ public void onSegmentChange(Segment newValue, ScoreboardModel.SegmentType segmen
+                QuestManager.getQuestFromName(questName.toString().trim());
+        if (questInfoOpt.isEmpty()) {
+            WynntilsMod.warn("Cannot match quest from scoreboard to actual quest: " + questName);
+            return;
```
Perhaps do something like 
```java
questInfo.setNextTask("Unknown");
```
if possible so the player knows that parsing has failed instead of showing an outdated task

magicus replied:
But if that happens we don't have a questInfo... If you're talking about what to do in case `questInfoOpt.isEmpty()`.

I think this error is unlikely enough that we don't have to worry about it. And it won't cause a crash.

Incompleteusern replied:
It's more that I'd rather we not show the user outdated quest info when parsing failed, though this is niche as you say.

Incompleteusern reviewed (dismissed):
A few minor comments, LGTM

magicus commented:
@Incompleteusern More or less all your comments applied as much to the code before I moved it. (The one exception is that I added the "play sound" line for opening on the wiki). 

But I'll go through them one by one and see what I can do.

magicus committed:
Clarify DialogueHistoryQueries

magicus committed:
Clarify difference between constant metadata and everchanging progress

magicus committed:
chore: spotless formatting

magicus committed:
Remove "sorted" from name

magicus committed:
chore: spotless formatting

magicus committed:
Fix wiki URLs

magicus committed:
chore: spotless formatting
